book/layout.md

```diff
@@ -295,56 +282,87 @@ Size and position
+greedy: paragraph and headings and other things take up all the
+horizontal space they can. So they use the parent's width:
 
 ``` {.python}
 self.w = self.parent.w
 ```
 
-Then, each of the children need to be laid out. That means each child
-has to be positioned, and its `layout` method must be called:
+This also means that a block layout object's horizontal position is
+the same as its parent's:
+
+``` {.python}
+self.x = self.parent.x
+```
+
+The vertical position of a block layout object depends on whether it
+is its parent's first child or not. First children start at the top of
+the parent, later children start where the previous layout object
+ended:
+
+``` {.python}
+if self.previous:
+    self.y = self.previous.y + self.previous.h
+else:
+    self.y = self.parent.y
+```
+
+After the width and position is computed, we can lay out all the
+children of a block layout:
 
 ``` {.python indent=8}
-y = self.y
 for child in self.children:
-    child.x = self.x
-    child.y = y
     child.layout()
-    y += child.h
 ```
 
 Note that the call to the child's `layout` method not only asks it to
 compute its size, but also to its own children to the layout tree and
 recursively call `layout` on them.
 
-The height of an element must include all its children. Since the
-children's heights have already all been summed into `y`, we can just
-subtract its starting value to compute the height:
+Finally, once all the children are laid out, a height of the layout
+object can be computed:
 
 ``` {.python}
-self.h = y - self.y
+self.h = sum([child.h for child in self.children])
 ```
 
-That settles `BlockLayout`; let's now work on `InlineLayout`. Its `x`
-and `y` will be set by its parent, so its `layout` just needs to set
-`w` and `h`:
+Note how important the order is here. A block layout's width depends
+on the parent's width; so by symmetry the width must be computed
+before laying out the children. The position is the same: it has to be
+computed before recursing so it's already available when the children
+compute their own positions. But the height works the opposite way:
+the height of a block layout depends on the height of its children, so
+it must be computed after recursing, so that the heights of the
+children are available.
+
+This *dependency ordering* plays a really important role in
+understanding how layout works, especially as layout gets more
+complicated to support more features and faster speed. [Chapter
+10](reflow.md) will explore this topic more.
+
+That settles `BlockLayout`; let's now work on `InlineLayout`. Its `x`,
+`w`, and `y` are set the same way as for a `BlockLayout`, but its
+height is computed based on its *y*-cursor `cy` instead of the height
+of its children.
 
 ``` {.python}
 class InlineLayout:
     def layout(self):
+        self.x = self.parent.x
         self.w = self.parent.w
+
+        if self.previous:
+            self.y = self.previous.y + self.previous.h
```

Comment:
Where did self.y get set in the old code? Don't see any line deletions in your PR for that.

Reply:
In the previous version, `x` and `y` were set by one's parent, and since `InlineLayout` doesn't have any children (at this stage of the book) there isn't code for `y` to delete. Honestly that was super confusing.

Reply:
Got it, makes sense. A nice improvement.